src/accessibility/LandmarkNavigation.ts

```diff
@@ -0,0 +1,138 @@
+        const nextIndex = LandmarkNavigation.getIndexInArray(currentIndex + offset);
+        const newLandmark = ORDERED_LANDMARKS[nextIndex];
+        return newLandmark;
+    }
+
+    /**
+     * Focus the next landmark from a given landmark.
+     * This method will skip over any missing landmarks.
+     * @param currentLandmark The current landmark
+     * @param backwards If true, search the next landmark to the left in ORDERED_LANDMARKS
+     */
+    private static findAndFocusNextLandmark(currentLandmark: Landmark, backwards = false): void {
+        let somethingWasFocused = false;
+        let landmark = currentLandmark;
+        while (!somethingWasFocused) {
+            landmark = LandmarkNavigation.getLandmark(landmark, backwards);
+            somethingWasFocused = focusFunctions[landmark]();
+        }
+    }
+
+    /**
+     * Find and focus the next landmark from a given landmark
+     * @param currentLandmark The current landmark
+     */
+    public static navigateToNextLandmarkFrom(currentLandmark: Landmark): void {
+        LandmarkNavigation.findAndFocusNextLandmark(currentLandmark);
+    }
+
+    /**
+     * Find and focus the previous landmark from a given landmark
+     * @param currentLandmark The current landmark
+     */
+    public static navigateToPreviousLandmarkFrom(currentLandmark: Landmark): void {
+        LandmarkNavigation.findAndFocusNextLandmark(currentLandmark, true);
+    }
+}
+
+/**
+ * Keys are the different landmarks and the values are function which when invoked
+ * gives focus to the corresponding landmark. These functions return a boolean
+ * indicating whether focus was successfully given.
+ */
+const focusFunctions = {
+    [Landmark.ACTIVE_SPACE_BUTTON]: () => {
+        const e = document.querySelector<HTMLElement>(".mx_SpaceButton_active");
+        e?.focus();
+        return !!e;
+    },
+    [Landmark.ROOM_SEARCH]: () => {
+        const e = document.querySelector<HTMLElement>(".mx_RoomSearch");
+        e?.focus();
+        return !!e;
+    },
+    [Landmark.ROOM_LIST]: () => {
+        const e =
+            document.querySelector<HTMLElement>(".mx_RoomTile_selected") ||
+            document.querySelector<HTMLElement>(".mx_RoomTile");
+        e?.focus();
+        return !!e;
+    },
+    [Landmark.MESSAGE_COMPOSER_OR_HOME]: () => {
+        const isComposerOpen = !!document.querySelector(".mx_MessageComposer");
+        if (isComposerOpen) {
+            const inThread = !!document.activeElement?.closest(".mx_ThreadView");
+            defaultDispatcher.dispatch(
+                {
+                    action: Action.FocusSendMessageComposer,
+                    context: inThread ? TimelineRenderingType.Thread : TimelineRenderingType.Room,
+                },
+                true,
+            );
+            return true;
+        } else {
+            const e = document.querySelector<HTMLElement>(".mx_HomePage");
+            e?.focus();
+            return !!e;
+        }
+    },
```

Comment:
Couldn't the signature be that it returns the `e` and the caller focuses it if `!undefined` otherwise moves onto the next one? would get rid of this 2 line boilerplate in every focusFunction

Reply:
https://github.com/matrix-org/matrix-react-sdk/pull/12190/commits/ea4fea8b96af0c010ce4598b04459d7bc3046549